mlflow/pyspark/ml/__init__.py

```diff
@@ -167,7 +249,86 @@ def _get_instance_param_map_recursively(instance, level):
+    # in the parameter search space was explored
+    child_run_start_time = parent_run.info.start_time
+    child_run_end_time = int(time.time() * 1000)
+
+    estimator_param_maps = parent_estimator.getEstimatorParamMaps()
+    tuning_estimator = parent_estimator.getEstimator()
+    metric_key = parent_estimator.getEvaluator().getMetricName()
+    if isinstance(parent_model, CrossValidatorModel):
+        metric_key = 'avg_' + metric_key
+        metrics = parent_model.avgMetrics
+    elif isinstance(parent_model, TrainValidationSplitModel):
+        metrics = parent_model.avgMetrics
+    else:
+        raise RuntimeError(f'Unknown parameter search model type {type(parent_model)}.')
+
+    for i in range(len(estimator_param_maps)):
+        estimator = tuning_estimator.copy(estimator_param_maps[i])
+        tags_to_log = dict(child_tags) if child_tags else {}
+        tags_to_log.update({MLFLOW_PARENT_RUN_ID: parent_run.info.run_id})
+        tags_to_log.update(_get_estimator_info_tags(estimator))
+
+        child_run = client.create_run(
+            experiment_id=parent_run.info.experiment_id,
+            start_time=child_run_start_time,
+            tags=tags_to_log,
+        )
+
+        params_to_log = _get_instance_param_map(estimator)
+        param_batches_to_log = _chunk_dict(params_to_log, chunk_size=MAX_PARAMS_TAGS_PER_BATCH)
+        metrics_to_log = {metric_key: metrics[i]}
+        for params_batch, metrics_batch in zip_longest(
+                param_batches_to_log, [metrics_to_log], fillvalue={}
+        ):
+            # Trim any parameter keys / values and metric keys that exceed the limits
+            # imposed by corresponding MLflow Tracking APIs (e.g., LogParam, LogMetric)
+            truncated_params_batch = _truncate_dict(
+                params_batch, MAX_ENTITY_KEY_LENGTH, MAX_PARAM_VAL_LENGTH
+            )
+            truncated_metrics_batch = _truncate_dict(
+                metrics_batch, max_key_length=MAX_ENTITY_KEY_LENGTH
+            )
+            client.log_batch(
+                run_id=child_run.info.run_id,
+                params=[
+                    Param(str(key), str(value)) for key, value in truncated_params_batch.items()
+                ],
+                metrics=[
+                    Metric(key=str(key), value=value, timestamp=child_run_end_time, step=0)
+                    for key, value in truncated_metrics_batch.items()
+                ],
+            )
+        client.set_terminated(run_id=child_run.info.run_id, end_time=child_run_end_time)
+
+
+def _log_parameter_search_results_as_artifact(param_maps, metrics, metric_name, run_id):
+    import pandas as pd
+    result_dict = {key: [] for key in param_maps[0].keys()}
+    result_dict[metric_name] = []
+    for i in range(len(param_maps)):
+        for param_name, param_value in param_maps[i].items():
+            result_dict[param_name].append(param_value)
+        result_dict[metric_name].append(metrics[i])
```

Comment:
I think this code can be simplifed to:

```python
from collections import defaultdict

result_dict = defaultdict(list)
result_dict[metric_name] = metrics
for param_map in param_maps:
    for param_name, param_value in param_map.items():
        result_dict[param_name].append(param_value)
```

Reply:
I re-check the sklearn side, it log a single column called "params" and that column include K-V values of all params. So I updated to be similar style.

Reply:
Have you checked how `cv_results.csv` looks like?

Reply:
<img width="404" alt="Screen Shot 2021-04-29 at 8 24 41" src="https://user-images.githubusercontent.com/17039389/116484662-7c748700-a8c4-11eb-8753-977d446744f6.png">

> it log a single column called "params"

sklearn autologging creates a parameter named `grid_search`. Are you talking about this parameter?

Reply:
Ah I just found `cv_results.csv` has a column named `params`:

![Screen Shot 2021-04-29 at 9 18 22](https://user-images.githubusercontent.com/17039389/116487844-e47a9b80-a8cb-11eb-9e07-948e14e9ab33.png)

Reply:
Updated. Add both single param column and param map column.